File: go/osv/models/internal/validate/validate.go
// Package main reads datastore records that were created by the validate.py python script.
// This should not be run outside of the validate.py script.
package main

import (
	"context"
	"fmt"
	"os"
	"time"

	"cloud.google.com/go/datastore"
	"github.com/google/osv.dev/go/osv/models"
)

func main() {
	ctx := context.Background()
	// Note: this does not communicate with GCP.
	// The Python code that runs this sets up the datastore emulator
	// with the relevant environment variables necessary.
	client, err := datastore.NewClient(ctx, os.Getenv("GOOGLE_CLOUD_PROJECT"))
	if err != nil {
		fmt.Printf("(Go) Failed creating Datastore client: %v\n", err)
		os.Exit(1)
	}

	readRecords(ctx, client)
	writeRecords(ctx, client)
}

func readRecords(ctx context.Context, client *datastore.Client) {
	fmt.Println("(Go) Getting Vulnerability")
	key := datastore.NameKey("Vulnerability", "CVE-123-456", nil)
	var vulnerability models.Vulnerability
	if err := client.Get(ctx, key, &vulnerability); err != nil {
		fmt.Printf("(Go) Failed getting Vulnerability: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("(Go) Getting AliasGroup")
	key = datastore.NameKey("AliasGroup", "1", nil)
	var aliasGroup models.AliasGroup
	if err := client.Get(ctx, key, &aliasGroup); err != nil {
		fmt.Printf("(Go) Failed getting AliasGroup: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("(Go) Getting AliasAllowListEntry")
	key = datastore.NameKey("AliasAllowListEntry", "1", nil)
	var aliasAllowListEntry models.AliasAllowListEntry
	if err := client.Get(ctx, key, &aliasAllowListEntry); err != nil {
		fmt.Printf("(Go) Failed getting AliasAllowListEntry: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("(Go) Getting AliasDenyListEntry")
	key = datastore.NameKey("AliasDenyListEntry", "1", nil)
	var aliasDenyListEntry models.AliasDenyListEntry
	if err := client.Get(ctx, key, &aliasDenyListEntry); err != nil {
		fmt.Printf("(Go) Failed getting AliasDenyListEntry: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("(Go) Getting UpstreamGroup")
	key = datastore.NameKey("UpstreamGroup", "1", nil)
	var upstreamGroup models.UpstreamGroup
	if err := client.Get(ctx, key, &upstreamGroup); err != nil {
		fmt.Printf("(Go) Failed getting UpstreamGroup: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("(Go) Getting ListedVulnerability")
	key = datastore.NameKey("ListedVulnerability", "CVE-123-456", nil)
	var listedVulnerability models.ListedVulnerability
	if err := client.Get(ctx, key, &listedVulnerability); err != nil {
		fmt.Printf("(Go) Failed getting ListedVulnerability: %v\n", err)
		os.Exit(1)
	}
}

func writeRecords(ctx context.Context, client *datastore.Client) {
	fmt.Println("(Go) Writing Vulnerability")
	key := datastore.NameKey("Vulnerability", "CVE-987-654", nil)
	vulnerability := models.Vulnerability{
		SourceID:    "test:path/to/CVE-987-654",
		Modified:    time.Date(2025, time.December, 31, 23, 59, 59, 0, time.UTC),
		IsWithdrawn: false,
		ModifiedRaw: time.Date(2025, time.December, 1, 23, 59, 59, 0, time.UTC),
		AliasRaw:    []string{"OSV-987-654", "TEST-987-654"},
		RelatedRaw:  []string{"CVE-999-999"},
		UpstreamRaw: []string{"CVE-987-000", "OSV-987-000"},
	}
	if _, err := client.Put(ctx, key, &vulnerability); err != nil {
		fmt.Printf("(Go) Failed writing Vulnerability %v: %v\n", key, err)
		os.Exit(1)
	}

	fmt.Println("(Go) Writing AliasGroup")
	key = datastore.NameKey("AliasGroup", "2", nil)
	aliasGroup := models.AliasGroup{
		VulnIDs:  []string{"A-1", "B-1", "C-1"},
		Modified: time.Date(2025, time.January, 1, 1, 1, 1, 1, time.UTC),
	}
	if _, err := client.Put(ctx, key, &aliasGroup); err != nil {
		fmt.Printf("(Go) Failed writing AliasGroup %v: %v\n", key, err)
		os.Exit(1)
	}

	fmt.Println("(Go) Writing AliasAllowListEntry")
	key = datastore.NameKey("AliasAllowListEntry", "2", nil)
	aliasAllowListEntry := models.AliasAllowListEntry{
		VulnID: "IS-GOOD",
	}
	if _, err := client.Put(ctx, key, &aliasAllowListEntry); err != nil {
		fmt.Printf("(Go) Failed writing AliasAllowListEntry %v: %v\n", key, err)
		os.Exit(1)
	}

	fmt.Println("(Go) Writing AliasDenyListEntry")
	key = datastore.NameKey("AliasDenyListEntry", "2", nil)
	aliasDenyListEntry := models.AliasDenyListEntry{
		VulnID: "IS-BAD",
	}
	if _, err := client.Put(ctx, key, &aliasDenyListEntry); err != nil {
		fmt.Printf("(Go) Failed writing AliasDenyListEntry %v: %v\n", key, err)
		os.Exit(1)
	}

	fmt.Println("(Go) Writing UpstreamGroup")
	key = datastore.NameKey("UpstreamGroup", "2", nil)
	upstreamGroup := models.UpstreamGroup{
		UpstreamIDs:       []string{"U-1", "U-2"},
		Modified:          time.Date(2025, time.January, 1, 1, 1, 1, 1, time.UTC),
		UpstreamHierarchy: []byte(`{"A": ["B"]}`),
	}
	if _, err := client.Put(ctx, key, &upstreamGroup); err != nil {
		fmt.Printf("(Go) Failed writing UpstreamGroup %v: %v\n", key, err)
		os.Exit(1)
	}

	fmt.Println("(Go) Writing ListedVulnerability")
	key = datastore.NameKey("ListedVulnerability", "CVE-987-654", nil)
	listedVulnerability := models.ListedVulnerability{
		Published:  time.Date(2025, time.December, 31, 23, 59, 59, 0, time.UTC),
		Ecosystems: []string{"Go", "PyPI"},
		Packages:   []string{"stdlib", "requests"},
		Summary:    "A vulnerability",
		IsFixed:    true,
		Severities: []models.Severity{
			{Type: "CVSS_V3", Score: "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H"},
		},
		AutocompleteTags: []string{"cve-987-654", "stdlib", "requests"},
		SearchIndices:    []string{"cve-987-654", "stdlib", "requests"},
	}
	if _, err := client.Put(ctx, key, &listedVulnerability); err != nil {
		fmt.Printf("(Go) Failed writing ListedVulnerability %v: %v\n", key, err)
		os.Exit(1)
	}
}

File: go/osv/models/internal/validate/validate.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Script to ensure datastore definitions in Python and Go are compatible."""
import datetime
import subprocess
import sys

from google.cloud import ndb

import osv.tests
from osv import Vulnerability, AliasGroup, AliasAllowListEntry, \
    AliasDenyListEntry, ListedVulnerability, Severity, UpstreamGroup, \
    RelatedGroup


def main() -> int:
  # Populate the examples from Python
  print('(Python) Putting Vulnerability')
  Vulnerability(
      id='CVE-123-456',
      source_id='test:path/to/CVE-123-456.json',
      modified=datetime.datetime(2025, 1, 2, 3, 4, 5, tzinfo=datetime.UTC),
      is_withdrawn=False,
      modified_raw=datetime.datetime(2025, 1, 1, 1, 1, 1, tzinfo=datetime.UTC),
      alias_raw=['OSV-123-456', 'TEST-123-456'],
      related_raw=['CVE-000-000', 'CVE-111-111'],
      upstream_raw=['CVE-123-000', 'OSV-123-000'],
  ).put()

  print('(Python) Putting AliasGroup')
  AliasGroup(
      id='1',
      bug_ids=['CVE-123-456', 'OSV-123-456', 'TEST-123-456'],
      last_modified=datetime.datetime(
          2025, 6, 7, 8, 9, 10, tzinfo=datetime.UTC),
  ).put()

  print('(Python) Putting AliasAllowListEntry')
  AliasAllowListEntry(
      id='1',
      bug_id='GOOD-VULN',
  ).put()

  print('(Python) Putting AlaisDenyListEntry')
  AliasDenyListEntry(
      id='1',
      bug_id='BAD-VULN',
  ).put()

  print('(Python) Putting UpstreamGroup')
  UpstreamGroup(
      id='1',
      upstream_ids=['U-1', 'U-2'],
      last_modified=datetime.datetime(
          2025, 6, 7, 8, 9, 10, tzinfo=datetime.UTC),
      upstream_hierarchy='{"A": ["B"]}',
  ).put()

  print('(Python) Putting ListedVulnerability')
  ListedVulnerability(
      id='CVE-123-456',
      published=datetime.datetime(2025, 1, 2, 3, 4, 5, tzinfo=datetime.UTC),
      ecosystems=['Go', 'PyPI'],
      packages=['stdlib', 'requests'],
      summary='A vulnerability',
      is_fixed=True,
      severities=[
          Severity(
              type='CVSS_V3',
              score='CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H')
      ],
      autocomplete_tags=['cve-123-456', 'stdlib', 'requests'],
      search_indices=['cve-123-456', 'stdlib', 'requests'],
  ).put()

  print('(Python) Putting RelatedGroup')
  RelatedGroup(
      id='CVE-123-456',
      related_ids=['R-1', 'R-2'],
      modified=datetime.datetime(2025, 6, 7, 8, 9, 10, tzinfo=datetime.UTC),
  ).put()

  # Run Go program to read the Python-created entities in Go.
  # And write Go entities.
  result = subprocess.run(['go', 'run', './validate.go'], check=False, cwd='.')
  if result.returncode != 0:
    return result.returncode

  # Read the Go-created entities in Python.
  print('(Python) Getting Vulnerability')
  if Vulnerability.get_by_id('CVE-987-654') is None:
    return 1
  print('(Python) Getting AliasGroup')
  if AliasGroup.get_by_id('2') is None:
    return 1
  print('(Python) Getting AliasAllowListEntry')
  if AliasAllowListEntry.get_by_id('2') is None:
    return 1
  print('(Python) Getting AliasDenyListEntry')
  if AliasDenyListEntry.get_by_id('2') is None:
    return 1
  print('(Python) Getting UpstreamGroup')
  if UpstreamGroup.get_by_id('2') is None:
    return 1
  print('(Python) Getting ListedVulnerability')
  if ListedVulnerability.get_by_id('CVE-987-654') is None:
    return 1
  print('(Python) Getting RelatedGroup')
  if RelatedGroup.get_by_id('CVE-987-654') is None:
    return 1

  return 0


if __name__ == '__main__':
  with osv.tests.datastore_emulator(), ndb.Client().context():
    ret = main()
  sys.exit(ret)
